app/tests/core_tests/test_tag_substitutions.py

```diff
@@ -0,0 +1,102 @@
+        ("foo", "bar", "[ bar ]", "[ bar ]"),
+        ("foo", "bar", "foo[\nfoo\n]", "foobar"),
+        ("foo", "bar", "[ foo<a> ]", "[ foo<a> ]"),
+        ("foo", "<b>bar</b>", "[ foo ]", "<b>bar</b>"),
+        ("foo", lambda: "bar", "[ foo ]", "bar"),
+    ],
+)
+def test_tag_substitution(tag_name, content, inp, out):
+    s = TagSubstitution(tag_name=tag_name, replacement=content)(inp)
+    assert s == out
+
+
+@pytest.mark.parametrize("tag_name", ["fo o", "fo\no", "'foo'", '"foo"'])
+def test_no_spaces(tag_name):
+    with pytest.raises(ValueError) as e:
+        TagSubstitution(tag_name=tag_name, replacement="blah")
+    assert "not a valid tag name" in str(e)
+
+
+@pytest.mark.parametrize(
+    "inp,output, repl",
+    [
+        ("[ foo 23aZ-_ ]", "bar23aZ-_", lambda a: f"bar{a}"),
+        ("[ foo 23 ]", "bar23", lambda a: f"bar{a}"),
+        ("[ foo '23' ]", "bar23", lambda a: f"bar{a}"),
+        ('[ foo "23" ]', "bar23", lambda a: f"bar{a}"),
+        ("[ foo 2 3 ]", "bar23", lambda a, b: f"bar{a}{b}"),
+        ("[ foo 2 3 4 ]", "bar234", lambda a, b, c: f"bar{a}{b}{c}"),
+    ],
+)
+def test_substitution_with_arguments(inp, output, repl):
+    s = TagSubstitution(
+        tag_name="foo",
+        replacement=repl,
+    )(inp)
+    assert s == output
+
+
+@pytest.mark.parametrize(
+    "inp,content,typ",
+    [
+        ("[ foo ]", "<a>", str),
+        (mark_safe("[ foo ]"), "<a>", str),
+        ("[ foo ]", mark_safe("<a>"), str),
+        (mark_safe("[ foo ]"), mark_safe("<a>"), SafeString),
+        ("[ foo ]", lambda: "<a>", str),
+        ("[ foo ]", lambda: mark_safe("<a>"), str),
+        (mark_safe("[ foo ]"), lambda: "<a>", str),
+        (mark_safe("[ foo ]"), lambda: mark_safe("<a>"), SafeString),
+    ],
+)
+def test_safe_substitutions(inp, content, typ):
+    result = TagSubstitution(tag_name="foo", replacement=content)(inp)
+    assert isinstance(result, typ)
+
+
+def _arg_dependent_replacement(arg):
+    if arg == "unsafe":
+        return "unsafe"
+    else:
+        return mark_safe("safe")
+
+
+@pytest.mark.parametrize(
+    "inp,content,typ",
+    [
+        ("[ foo unsafe ]", _arg_dependent_replacement, str),
+        (mark_safe("[ foo unsafe ]"), _arg_dependent_replacement, str),
+        ("[ foo safe ]", _arg_dependent_replacement, str),
+        (mark_safe("[ foo safe ]"), _arg_dependent_replacement, SafeString),
+        (mark_safe("[ foo unsafe ]"), _arg_dependent_replacement, str),
+        # multiple substitutions of mixed safety
+        (
+            mark_safe("[ foo unsafe ] [ foo safe ]"),
+            _arg_dependent_replacement,
+            str,
+        ),
+        (
+            mark_safe("[ foo safe ] [ foo safe ]"),
+            _arg_dependent_replacement,
+            SafeString,
+        ),
+    ],
+)
+def test_safe_substitutions_with_arguments(inp, content, typ):
+    result = TagSubstitution(tag_name="foo", replacement=content)(inp)
+    assert isinstance(result, typ)
```

Comment:
What is this test doing? I do not really understand what is going on with `_arg_dependent_replacement`. Why is it being added as a parameter anyway? Seems to be the same in all tests.

Reply:
Did you end up changing it? I am also not sure why past Chris didn't just add it as fixed.